scp/lib/bfs/bucket/scp.rb

```diff
@@ -145,32 +145,53 @@ def mkdir_p(path)
       end
 
       def sh!(*cmd) # rubocop:disable Metrics/MethodLength
-        stdout = ''
         stderr = nil
         status = 0
         cmdstr = cmd.map {|x| Shellwords.escape(x) }.join(' ')
 
+        # init command's STDOUT stream (large output may break lines in on_data):
+        stdout_reader, stdout_writer = IO.pipe
+
         @client.session.open_channel do |ch|
           ch.exec(cmdstr) do |_, _success|
             ch.on_data do |_, data|
-              if block_given?
-                yield data
-              else
-                stdout += data
-              end
+              stdout_writer.write(data)
             end
             ch.on_extended_data do |_, _, data|
               stderr = data
             end
             ch.on_request('exit-status') do |_, buf|
               status = buf.read_long
             end
+            ch.on_close do |_|
+              stdout_writer.close
+            end
           end
         end
-        @client.session.loop
+
+        # loop in background (blocking to be done by consumer):
+        loop_thread = Thread.new { @client.session.loop }
+
+        # consume command's STDOUT (blocking):
+        stdout = ''
+        if block_given?
+          # pass command's STDOUT to caller for reading (slurping or iterating lines etc):
+          yield stdout_reader
+        else
+          # or just slurp command's STDOUT into retval (`read(length=nil)` == "read until EOF"):
+          stdout = stdout_reader.read
+        end
+
+        # wait for completion:
+        loop_thread.join
+
+        # raise error if command completed with error:
         raise CommandError.new(cmdstr, status, stderr) unless status.zero?
 
         stdout
+      ensure
+        stdout_writer&.close
+        stdout_reader&.close
```

Comment:
You can work with pipes using block (which ensures that stuff is all closed):

```ruby
IO.pipe do |reader, writer|
  ...
end
```

But I thought it'll be too deep nesting, so decided to follow `reader, writer = IO.pipe` approach.